projects/plutus/src/current_sense.c

```diff
@@ -0,0 +1,90 @@
+#include "current_sense.h"
+
+#include <string.h>
+
+#include "critical_section.h"
+#include "wait.h"
+
+static void prv_calculate_current(int32_t *value, void *context) {
+  CurrentSenseStorage *storage = (CurrentSenseStorage *)context;
+
+  // Update the offset if the flag is set
+  if (storage->offset_flag) {
+    storage->offset = *value - storage->data.zero_point.voltage;
+    storage->offset_flag = false;
+  }
+
+  // Formula for calculating calibrated current. Draws slope between given calibrated
+  // points, and uses the result as well as the voltage offset to calculate current
+  storage->value.voltage = *value;
+  storage->value.current = (*value - storage->data.zero_point.voltage - storage->offset) *
+                           ((storage->data.max_point.current - storage->data.zero_point.current) /
+                            (storage->data.max_point.voltage - storage->data.zero_point.voltage));
+
+  if (storage->callback != NULL) {
+    storage->callback(storage->value.current, storage->context);
+  }
+}
+
+StatusCode current_sense_init(CurrentSenseStorage *storage, const CurrentSenseCalibrationData data,
+                              const LtcAdcSettings *settings) {
+  if (storage == NULL || settings == NULL) {
+    return status_code(STATUS_CODE_UNINITIALIZED);
+  }
+
+  status_ok_or_return(ltc_adc_init(&storage->adc_storage, settings));
+
+  // Store calibration parameters
+  storage->data = data;
```

Comment:
Can you store a copy of the calibration data or make it more obvious that `data` should persist?

Reply:
`data` is already being copied. It was changed in the storage structure from a pointer to an actual instance of `CurrentSenseValue`

Reply:
Oh, why did you make `data` pass by value instead of pass by reference? There's no point in making a pass by value const. You should make `data` a pointer.